src/pp/if_parser.rs

```diff
@@ -0,0 +1,436 @@
+                    Some(t) => Some(t),
+                    None => self.next()?,
+                }
+            }
+            _ => Some(token),
+        })
+    }
+
+    fn peek(&mut self) -> Step<Option<Token>> {
+        let token = match self.raw_peek() {
+            Some(t) => t,
+            None => return Ok(None),
+        };
+
+        Ok(match token.value {
+            TokenValue::Ident(ref name) if name != "defined" => {
+                self.raw_next();
+
+                match self.add_define(name, token.location)? {
+                    Some(t) => {
+                        self.carry = Some(t);
+                        self.carry.clone()
+                    }
+                    None => self.peek()?,
+                }
+            }
+            _ => Some(token),
+        })
+    }
+
+    fn expect_raw_next(&mut self) -> Step<Token> {
+        let val = self.raw_next();
+        val.ok_or(StepExit::Error((
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn expect_next(&mut self) -> Step<Token> {
+        let val = self.next()?;
+        val.ok_or(StepExit::Error((
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn expect_peek(&mut self) -> Step<Token> {
+        let val = self.peek()?;
+        val.ok_or(StepExit::Error((
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn add_define(&mut self, name: &str, location: Location) -> Step<Option<Token>> {
+        if self
+            .macro_processor
+            .start_define_invocation(name, location, &mut self.lexer)?
+        {
+            Ok(None)
+        } else if self.parsing_if {
+            Ok(Some(Token {
+                value: TokenValue::Integer(Integer {
+                    value: 0,
+                    signed: true,
+                    width: 64,
+                }),
+                location,
+            }))
+        } else {
+            Err(StepExit::Error((
+                PreprocessorError::UnexpectedToken(TokenValue::Ident(name.to_string())),
+                location,
+            )))
+        }
+    }
+
+    fn handle_defined(&mut self) -> Step<i64> {
+        let next = self.expect_raw_next()?;
+
+        match next.value {
+            TokenValue::Ident(ref name) => Ok(self.lexer.defines.get(name).is_some() as i64),
+            TokenValue::Punct(Punct::LeftParen) => {
+                let name_token = self.expect_raw_next()?;
+                let name = match name_token.value {
+                    TokenValue::Ident(name) => Ok(name),
+                    value => Err(StepExit::Error((
+                        PreprocessorError::UnexpectedToken(value),
+                        name_token.location,
+                    ))),
+                }?;
+
+                let close_brace = self.expect_next()?;
+
+                match close_brace.value {
```

Comment:
We have many pieces of code like this, maybe we could have helpers like `DirectiveProcessor::expect_lexer_token` and `expect_lexer_ident`.

Reply:
Do you mean `MacroProcessor` because the if parser doesn't have access to the `DirectiveProcessor`

Reply:
Sorry this was unclear: I meant copy pasting similar code into `IfParser` ^^